File: LinkedList/singlylinkedlist.py

class Node:

  def __init__(self, value=None, next=None):
    self.data = value
    self.next = next


  def get_value(self):
    return self.data


  def get_next(self):
    return self.next


  def set_value(self, value):
    self.data = value


  def set_next(self, next):
    self.next = next



class SinglyLinkedList:

  def __init__(self):
    self.head = None


  def is_empty(self):
    return self.head == None


  def insert(self, value):
    self.head = Node(value, self.head)


  def length(self):
    curr  = self.head
    count = 0

    while curr != None:
      count += 1
      curr = curr.get_next()

    return count


  def contains(self, value):
    curr  = self.head
    found = False

    while curr != None and found == False:
      if curr.get_value() == value:
        found = True
      else:
        curr = curr.get_next()

    return found


  def remove(self, value):
    curr = self.head
    prev = None
    found = False

    while found == False:
      if curr.get_value() == value:
        found = True
      else:
        prev = curr
        curr = curr.get_next()

    if prev == None:
      self.head = curr.get_next()
    else:
      prev.set_next(curr.get_next())


  def reverse(self):
    curr = self.head
    last = None

    while curr != None:
      nxt = curr.get_next()
      curr.set_next(last)
      last = curr
      curr = nxt

    self.head = last


  def display(self):
    curr = self.head
    data = []

    while curr != None:
      data.append(curr.get_value())
      curr = curr.get_next()

    print data



if __name__ == '__main__':

  ll = SinglyLinkedList()
  ll.insert(3)
  ll.insert(2)
  ll.insert(42)
  ll.insert(4)

  print 'Length:', ll.length()
  print 'Searching 42:', ll.contains(42)
  print 'Removing 42:', ll.remove(42)
  print 'Searching 42:', ll.contains(42)

  ll.display()
  print 'Reversing:', ll.reverse()
  ll.display()

File: LinkedList/singlylinkedlist.c
#include <stdio.h>
#include <stdlib.h> 


struct Node {
  int value;
  struct Node *next;
};

struct SinglyLinkedList {
  int size;
  struct Node *head;
};


typedef struct Node Node;
typedef struct SinglyLinkedList SinglyLinkedList;


void init(SinglyLinkedList *);
void insert(SinglyLinkedList *, int value);
void is_empty(SinglyLinkedList *);
void reverse(SinglyLinkedList *);
int  delete(SinglyLinkedList *, int value);
int  contains(SinglyLinkedList *, int value);


void init(SinglyLinkedList *list) {
  list->head = NULL;
  list->size = 0;
}


void insert(SinglyLinkedList *list, int value) {
  if(list->head != NULL) {
    Node *temp = malloc(sizeof(Node));
    temp->value = value;
    temp->next = list->head;
    list->head = temp;
    list->size++;
  } else {
    list->head = malloc(sizeof(Node));
    list->head->value = value;
    list->size++;
  }
}


int contains(SinglyLinkedList *list, int value) {
  Node *curr = list->head;
  int found = 0;

  while(!found && curr != NULL) {
    if(curr->value == value) {
      found = 1;
    } else {
      curr = curr->next;
    }
  }

  return found;
}


int delete(SinglyLinkedList *list, int value) {
  Node *curr = list->head;
  Node *prev = NULL;
  int found  = 0;

  while(found == 0) {
    if(curr->value == value) {
      list->size--;
      found = 1;
    } else {
      prev = curr;
      curr = curr->next;
    }
  }

  if(prev == NULL) {
    list->head = curr->next;
  } else {
    prev->next = curr->next;
  }

  return found;
}


void reverse(SinglyLinkedList *list) {
  Node *curr = list->head;
  Node *last = NULL;

  while(curr != NULL) {
    Node *temp = curr->next;
    curr->next = last;
    last = curr;
    curr = temp;
  }

  list->head = last;
}



void main() {

  SinglyLinkedList *list = malloc(sizeof(SinglyLinkedList));

  insert(list, 2);
  insert(list, 3);
  insert(list, 42);
  insert(list, 4);

  printf("List size: %d\n", list->size);
  printf("Searching for 42: %d\n", contains(list, 42));
  printf("Removing 42: %d\n", delete(list, 42));
  printf("Searching for 42: %d\n", contains(list, 42));

  printf("Getting first value: %d\n", list->head->value);

  printf("Reversing\n");
  reverse(list);

  printf("Getting first value: %d\n", list->head->value);

}
